src/Functions/FunctionBase58Conversion.h

```diff
@@ -0,0 +1,226 @@
+#pragma once
+#include "config_functions.h"
+
+#if USE_BASE58
+#    include <Columns/ColumnConst.h>
+#    include <Common/MemorySanitizer.h>
+#    include <Columns/ColumnString.h>
+#    include <DataTypes/DataTypeString.h>
+#    include <Functions/FunctionFactory.h>
+#    include <Functions/FunctionHelpers.h>
+#    include <IO/WriteHelpers.h>
+#    include <base_x.hh>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int BAD_ARGUMENTS;
+}
+
+struct Base58Encode
+{
+    static constexpr auto name = "base58Encode";
+
+    static void process(const ColumnString * input, ColumnString::MutablePtr& dst_column, std::string& alphabet, size_t input_rows_count)
+    {
+        auto & dst_data = dst_column->getChars();
+        auto & dst_offsets = dst_column->getOffsets();
+
+        size_t current_allocated_size = input->getChars().size();
+
+        dst_data.resize(current_allocated_size);
+        dst_offsets.resize(input_rows_count);
+
+        const ColumnString::Offsets & src_offsets = input->getOffsets();
+
+        const auto * source = input->getChars().raw_data();
+        auto * dst = dst_data.data();
+        auto * dst_pos = dst;
+
+        size_t src_offset_prev = 0;
+        size_t processed_size = 0;
+
+        const auto& encoder = (alphabet == "bitcoin") ? Base58::bitcoin() :
+                             ((alphabet == "flickr") ? Base58::flickr() :
+                             ((alphabet == "ripple") ? Base58::ripple() :
+                                                       Base58::base58()));
+
+        for (size_t row = 0; row < input_rows_count; ++row)
+        {
+            size_t srclen = src_offsets[row] - src_offset_prev - 1;
+            /// Why we didn't use char* here?
+            /// We don't know the size of the result string beforehand (it's not byte-to-byte encoding),
+            /// so we may need to do many resizes (the worst case -- we'll do it for each row)
+            /// This way we do exponential resizes and one final resize after whole operation is complete
+            std::string encoded;
+            if (srclen)
+                encoder.encode(encoded, source, srclen);
```

Comment:
Excessive copy and possibly, allocation in the loop.
We can instead make a string-like object that will write data directly into the column. See read*StringInto in ReadHelpers.h

Reply:
Although base58 is already very slow, maybe excessive copying will be unnoticeable.

Reply:
> possibly, allocation in the loop.

We will need this anyway -- we don't know the required size beforehand, result size is undetermined. So I think the best way here is to allocate exponentially and make one final adjust. We can use heuristic estimation here, but allocations will still happen

Reply:
We can resize the column buffer directly - see the usages of
```
template <typename Vector>
void readEscapedStringInto(Vector & s, ReadBuffer & buf);
```

Reply:
Yes, we can
I suppose you mean using result column buffer as a buffer for writing the result of the encoder&decoder itself. 
But again, we'll need to determine its size proactively, which we cannot do. In our case `encode()` is an external function we have almost no access to, and we'll need to reallocate from inside the function. 
E.g. for `base64` we could allocate required space in advance and then operate on `char*` w/o unnecessary strings, but for `base58` we'll need a string or vector or smth like this to dynamically insert (and possibly reallocate) into it. However, the created string will be destroyed afterwards. 
The only thing we can avoid here -- creating multiple strings. I.e. I can create a `std::string` before the loop and reuse it. This will probably do a bit of speed-up :smirk:

Anyway, maybe I just didn't get your idea at all :confused: 
If so, I would be very grateful for more detailed explanation

Reply:
Base58 is using something like `template <typename String>`, you can pass your own type there.

Reply:
And also, you can determine the upper bound for the size in advance.

Reply:
I think I got what you mean.
Though I already moved to re-using the same object, but still we use `memcpy`.
Of course it slows the process down, but (as you noted previously) the base58 module itself is (really) not fast (and all other modules I found on GH work the same way), and there is not too much to copy, so I don't think this is really worth fighting for. 
If you insist on refactoring this part, I still can do it, it's just a matter of time spent on it vs speed increase it will give :smirk:

> you can determine the upper bound in advance

I'm not sure if I really know all cases for base58 encoding, but I think setting a strict limit for upper length of processed string is hardly a good idea

Reply:
The question is what the worst case upper limit looks like. If it's <1.5 x input.size() then pre-allocation might be the best option (and my understanding is that we are in this ballpark). If it's significantly bigger or is very data-dependent (e.g fluctuates betwen 1.0x and 10x of the input size) then implicit exponential resizes are a better alternative.